### PR TITLE
[WIP] Add a command to decommission a specific worker

### PR DESCRIPTION
**Pull request by YichuanSun**
### What changes are proposed in this pull request?

A new command to decommission a single worker.

### Why are the changes needed?

https://github.com/Alluxio/alluxio/issues/15343
https://github.com/Alluxio/alluxio/pull/14000

Thanks for PR 14000, we want to give a more detailed version.

### Does this PR introduce any user facing changes?

Yes, I'll add them in the future.

Please list the user-facing changes introduced by your change, including
  1. change in user-facing APIs
  2. addition or removal of property keys
  3. webui

design doc https://docs.google.com/document/d/1GDVghn_9AOViOoSDvt1sqR8tSdyUmZ6D8xKN9YmhRu0/edit#

### CODE REVIEW

**Comment by qian0817** [comment]
It looks like this PR has some duplication with the #15949 and #15956，Can you take a look and give some advice?

File: core/transport/src/main/proto/grpc/file_system_master.proto [hunk 1]
```diff
@@ -169,6 +169,18 @@ message FreePRequest {
   optional FreePOptions options = 2;
 }
 
+message FreeWorkerPResponse {}
+message FreeWorkerPOptions {
+  optional bool forced = 1;
+  optional int64 timeOut = 2;
+  optional FileSystemMasterCommonPOptions commonOption = 3;
```
Comment (qian0817) [inline core/transport/src/main/proto/grpc/file_system_master.proto:176]: `FileSystemMasterCommonPOptions` is useless and I recommend moving this RPC to `block_master.proto`.

**Review by qian0817 — CHANGES_REQUESTED** [review]

**Comment by YichuanSun** [comment]
Thanks for advice of @qian0817😁👍! Today I have no time to modify more codes, I will deal with your comments this morning.

**Comment by jiacheliu3** [comment]
@dbw9580 @YichuanSun PTAL there's another existing PR with a similar ask (but not identical): https://github.com/Alluxio/alluxio/pull/15155
Just wanted to make sure we don't end up having multiple commands with overlapping functionality.

**Review by jiacheliu3 — COMMENTED** [review]
Good work! I left a few high level comments.

File: core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 2]
```diff
@@ -392,6 +393,9 @@ boolean freeWorker(String workerName, FreeWorkerContext freeWorkerContext)
 
   void decommissionToFree(String workerName) throws UnavailableException;
 
+  void decommissionWorker(String workerName, DecommissionWorkerContext decommissionWorkerContext)
```
[the remark this answers is not in the document]
Comment (YichuanSun) [reply core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java:396]: Agreed. I confused the abstraction of BlockMaster and FileSystemMaster. I will modify them tomorrow. Thanks!

[the remark this answers is not in the document]
Comment (jiacheliu3) [reply core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java:396]: The BlockMaster manages blocks therefore also manages all worker interactions. The FileSystemMaster manages the file abstraction including file system interface, inodes, and the inode tree structure. I guess that's how you can distinguish the two.

**Comment by YichuanSun** [comment]
After rerethinking😅, the reboot method in AlluxioWorkerProcess is not thread-safe, I'll correct it.

File: core/server/worker/src/main/java/alluxio/worker/AlluxioWorkerProcess.java [hunk 3]
```diff
@@ -335,4 +344,35 @@ public WorkerNetAddress getAddress() {
   public String toString() {
     return "Alluxio worker @" + mRpcConnectAddress;
   }
+
+  public boolean rebootDataServer() throws IOException {
```
Comment (qian0817) [inline core/server/worker/src/main/java/alluxio/worker/AlluxioWorkerProcess.java:348]: why need  to reboot data sever?

**Comment by qian0817** [comment]
Is there design doc for this feature, for some of the implemented code I don't understand why.

**Comment by YichuanSun** [comment]
> Is there design doc for this feature, for some of the implemented code I don't understand why.

https://docs.google.com/document/d/1GDVghn_9AOViOoSDvt1sqR8tSdyUmZ6D8xKN9YmhRu0/edit#

File: core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (bzheng888) [reply core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java]: @jiacheliu3 Little confused, the BlockMaster has not a object member filesystemmaster, but the filesystemmaster has a member object blockmaster, so maybe let the logic in filesystemmaster easier 🤔️?

**Comment by YichuanSun** [comment]
> This pull request has been automatically marked as stale because it has not had recent activity. It will be closed in two weeks if no further activity occurs. Thank you for your contributions.

Do not stale it, I will improve this feature in the future.

**Pull request closed without merging**